common/lib/dependabot/dependency_updaters/README.md

```diff
@@ -0,0 +1,63 @@
+# Dependency Updater
+
+A *dependency updater* is responsible for updating a dependent package
+according to a set of requirements.
+It provides a minimal interface for languages to add Dependabot support
+by delegating responsibilities to a package manager for that language.
+
+## Public API
+
+Each `Dependabot::DependencyUpdaters` class inherits from `Base` 
+and implements the following methods:
+
+### `#update(dependency:, requirements:)`
+
+Updates a dependency according to the provided requirements.
+This method is called within the root directory of a project.
+
+#### Arguments
+
+| Argument       | Type                     | Description                      |
+| -------------- | ------------------------ | -------------------------------- |
+| `dependency`   | `Dependabot::Dependency` | The package dependency to update |
+| `requirements` | `VersionRange`           | The updated version constraints  |
```

Comment:
For security updates, would we pass in the inverted `affected_versions` of the advisory?

Reply:
Maybe. It depends on which perspective we want to favor. The advisory database gives us ranges affected versions, but package managers express positive requirement.

Any range or enumeration of affected versions can be trivially inverted, so it may not be an issue. I'd be interested to hear what implementors have to say once we have better primitives for requirements.

Reply:
> Any range or enumeration of affected versions can be trivially inverted, so it may not be an issue. I'd be interested to hear what implementors have to say once we have better primitives for requirements.

👍 

I think we'll also need a way to pass in ignored ranges from the [config file](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/configuration-options-for-dependency-updates#ignore)

Reply:
What do you think about passing a `Configuration` object as an argument?

Reply:
I like it, that also gives us an opportunity to pass in things like credentials for registries.

Even though we don't pass credentials in directly when running dependabot at GitHub (we run a proxy that adds auth details to requests), downstream users of the library often do depend on being able to pass the credentials into dependabot-core and having core make authenticated requests that way.

Reply:
I just sketched this out in 8a0a19b. Take a look and let me know what you think 😄